### PR TITLE
feat(storage): add support for IBM cloud object storage as storage client

### PR DESCRIPTION
**Pull request by athira-varghese**
**What this PR does / why we need it**:
   Add support for IBM cloud object storage as storage client
   
**Which issue(s) this PR fixes**:
Fixes NA

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated

### CODE REVIEW

**Review by MasslessParticle — APPROVED** [review]

**Comment by MasslessParticle** [comment]
This looks great, thanks for contributing! 

Please fix the conflicts and we'll be happy to merge

**Comment by adityacs** [comment]
@MasslessParticle We have resolved the conflicts and also fixed lint errors

File: docs/sources/configuration/_index.md [hunk 1]
```diff
@@ -4209,6 +4222,77 @@ The `swift_storage_config` block configures the connection to OpenStack Object S
...
+
+# COS Endpoint to connect to.
+# CLI flag: -common.storage.cos.endpoint
+[endpoint: <string> | default = ""]
+
+# COS region to use.
+# CLI flag: -common.storage.cos.region
+[region: <string> | default = ""]
+
+# COS HMAC Access Key ID
```
Comment (JStickler) [inline docs/sources/configuration/_index.md:4251]:
```suggestion
# COS HMAC Access Key ID.
```

File: docs/sources/configuration/_index.md [hunk 2]
```diff
@@ -4209,6 +4222,77 @@ The `swift_storage_config` block configures the connection to OpenStack Object S
...
+# CLI flag: -common.storage.cos.access-key-id
+[access_key_id: <string> | default = ""]
+
+# COS HMAC Secret Access Key
```
Comment (JStickler) [inline docs/sources/configuration/_index.md:4255]:
```suggestion
# COS HMAC Secret Access Key.
```

File: docs/sources/configuration/_index.md [hunk 3]
```diff
@@ -4209,6 +4222,77 @@ The `swift_storage_config` block configures the connection to OpenStack Object S
...
+  [idle_conn_timeout: <duration> | default = 1m30s]
+
+  # If non-zero, specifies the amount of time to wait for a server's response
+  # headers after fully writing the request.
+  # CLI flag: -common.storage.cos.http.response-header-timeout
+  [response_header_timeout: <duration> | default = 0s]
+
+# Configures back off when cos get Object.
+backoff_config:
+  # Minimum backoff time when cos get Object
```
Comment (JStickler) [inline docs/sources/configuration/_index.md:4271]:
```suggestion
  # Minimum backoff time when cos get Object.
```

File: docs/sources/configuration/_index.md [hunk 4]
```diff
@@ -4209,6 +4222,77 @@ The `swift_storage_config` block configures the connection to OpenStack Object S
...
+
+  # Maximum backoff time when cos get Object
+  # CLI flag: -common.storage.cos.max-backoff
+  [max_period: <duration> | default = 3s]
+
+  # Maximum number of times to retry when cos get Object
+  # CLI flag: -common.storage.cos.max-retries
+  [max_retries: <int> | default = 5]
+
+# api Key
```
Comment (JStickler) [inline docs/sources/configuration/_index.md:4283]: Should this be
```suggestion
# API key
```

File: docs/sources/configuration/_index.md [hunk 5]
```diff
@@ -4209,6 +4222,77 @@ The `swift_storage_config` block configures the connection to OpenStack Object S
...
+# CLI flag: -common.storage.cos.api-key
+[api_key: <string> | default = ""]
+
+# COS service instance id to use
```
Comment (JStickler) [inline docs/sources/configuration/_index.md:4287]:
```suggestion
# COS service instance id to use.
```

**Review by JStickler — COMMENTED** [review]
[docs team]  Couple of small suggestions.

**Comment by tareqmamari** [comment]
@JStickler we have addressed the docs related comments.

**Review by JStickler — APPROVED** [review]
[docs team] LGTM

**Comment by tareqmamari** [comment]
@JStickler @MasslessParticle since this PR has been approved, is there anything else we should do to get it merged? Thank you!

**Pull request merged**